test/testdrive/subscribe-output.td

```diff
@@ -0,0 +1,255 @@
+> FETCH ALL c
+<TIMESTAMP> 1 1 2 3
+<TIMESTAMP> 1 4 0 -4
+<TIMESTAMP> 1 2 5 -6
+> COMMIT
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t) WITHIN TIMESTAMP ORDER BY a, b, c
+> FETCH 1 c
+<TIMESTAMP> 1 1 2 3
+> FETCH 1 c
+<TIMESTAMP> 1 2 5 -6
+> FETCH 1 c
+<TIMESTAMP> 1 4 0 -4
+> COMMIT
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE t WITHIN TIMESTAMP ORDER BY a, b, c
+> FETCH 1 c
+<TIMESTAMP> 1 1 2 3
+> FETCH 1 c
+<TIMESTAMP> 1 2 5 -6
+> FETCH 1 c
+<TIMESTAMP> 1 4 0 -4
+> COMMIT
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t) WITHIN TIMESTAMP ORDER BY c, b, mz_diff, a
+> FETCH 1 c
+<TIMESTAMP> 1 2 5 -6
+> FETCH 1 c
+<TIMESTAMP> 1 4 0 -4
+> FETCH 1 c
+<TIMESTAMP> 1 1 2 3
+> COMMIT
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE t WITHIN TIMESTAMP ORDER BY c, b, mz_diff, a
+> FETCH 1 c
+<TIMESTAMP> 1 2 5 -6
+> FETCH 1 c
+<TIMESTAMP> 1 4 0 -4
+> FETCH 1 c
+<TIMESTAMP> 1 1 2 3
+> COMMIT
+
+> CREATE TABLE t0(key int, value int)
+> INSERT INTO t0 VALUES (1, 2), (2, 3)
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t0) WITHIN TIMESTAMP ORDER BY key, mz_diff
+> FETCH 1 c
+<TIMESTAMP> 1 1 2
+
+> FETCH 1 c
+<TIMESTAMP> 1 2 3
+
+$ postgres-execute connection=mz_system
+UPDATE t0 SET value = 4 WHERE key = 1
+
+> FETCH 1 c
+<TIMESTAMP> -1 1 2
+
+> FETCH 1 c
+<TIMESTAMP> 1 1 4
+> COMMIT
+
+# make sure progress messages are ok
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE t WITH (PROGRESS, SNAPSHOT) WITHIN TIMESTAMP ORDER BY c, b, a
+> FETCH 1 c
+<TIMESTAMP> true <null> <null> <null> <null>
+> FETCH ALL c
+<TIMESTAMP> false 1 2 5 -6
+<TIMESTAMP> false 1 4 0 -4
+<TIMESTAMP> false 1 1 2 3
+<TIMESTAMP> true <null> <null> <null> <null>
+> COMMIT
+
+
+! SUBSCRIBE (SELECT 1 as a, 2 as b union select 2, 1) WITHIN TIMESTAMP ORDER BY c
+contains: column "c" does not exist
+
+! SUBSCRIBE (SELECT 1 as c, 2 as c union select 2, 1) WITHIN TIMESTAMP ORDER BY c
+contains: column reference "c" is ambiguous
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT 1 as a, 2 as b union select 2, 1) ENVELOPE UPSERT (KEY (a, b))
+> FETCH 1 c
+<TIMESTAMP> upsert 1 2
+
+> FETCH 1 c
+<TIMESTAMP> upsert 2 1
+> COMMIT
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT 1 as a, 2 as b union select 2, 1) ENVELOPE UPSERT (KEY (b, a))
+> FETCH 1 c
+<TIMESTAMP> upsert 1 2
+
+> FETCH 1 c
+<TIMESTAMP> upsert 2 1
+> COMMIT
+
+> CREATE TABLE t1(key int, value int, UNIQUE (key))
+> INSERT INTO t1 VALUES (1, 2), (2, 3)
+
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t1) ENVELOPE UPSERT (KEY (key))
+> FETCH 2 c
+<TIMESTAMP> upsert 1 2
+<TIMESTAMP> upsert 2 3
+
+$ postgres-execute connection=mz_system
+UPDATE t1 SET value = 4 WHERE key = 1
+
+> FETCH 1 c
+<TIMESTAMP> upsert 1 4
+
+$ postgres-execute connection=mz_system
+DELETE FROM t1 WHERE key = 1
+
+> FETCH ALL c
+<TIMESTAMP> delete 1 <null>
+
+$ postgres-execute connection=mz_system
+INSERT INTO t1 VALUES (1, 1), (1, 1)
+
+> FETCH ALL c
+<TIMESTAMP> "key violation" 1 <null>
```

Comment:
I am afraid the presence of this "key violation" marker makes SUBSCRIBE behave differently as compared to an equivalent SINK. Consider this case:

```
DROP TABLE IF EXISTS t1 CASCADE;
CREATE TABLE t1 (f1 INTEGER, f2 INTEGER, PRIMARY KEY (f1));
INSERT INTO t1 VALUES (1,1), (1,2);
-- COPY (SUBSCRIBE t1 ENVELOPE UPSERT (KEY (f1))) TO STDOUT;
CREATE SINK s1 FROM t1
INTO KAFKA CONNECTION kafka_conn (TOPIC 'upsert-error-sink-td')
KEY (f1)
FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
ENVELOPE UPSERT;
```
The sink will happily process such input and output the following:

```
CreateTime:1682668844471	{"f1":1,"f2":{"int":1}}
CreateTime:1682668844471	{"f1":1,"f2":{"int":2}}
```

That is, two updates within the same timestamp. Therefore, if I started using Mz with sinks but converted to Subscribe for some reason, or the other way around, I would get different results for what is ostensibly the same `ENVELOPE`.

Reply:
Yeah, I agree with you but I think they have to be different. There was some discussion in slack about this too: https://materializeinc.slack.com/archives/C02FWJ94HME/p1682562893376109?thread_ts=1682534093.291879&cid=C02FWJ94HME